File: Lista/Java/Q15.java
class SaldoInsuficienteException extends Exception {
    public SaldoInsuficienteException(String mensagem) {
        super(mensagem);
    }
}

class ContaBancaria {
    private String titular;
    private double saldo;

    public ContaBancaria(String titular, double saldoInicial) {
        this.titular = titular;
        this.saldo = saldoInicial;
    }

    public void depositar(double valor) {
        if (valor > 0) {
            saldo += valor;
            System.out.println("Depósito de R$ " + valor + " realizado com sucesso.");
        } else {
            System.out.println("Valor de depósito deve ser positivo.");
        }
    }

    public void sacar(double valor) throws SaldoInsuficienteException {
        if (valor > saldo) {
            throw new SaldoInsuficienteException("Saldo insuficiente! Saldo atual: R$ " + saldo);
        } else {
            saldo -= valor;
            System.out.println("Saque de R$ " + valor + " realizado com sucesso.");
        }
    }

    public double getSaldo() {
        return saldo;
    }

    public void exibirInformacoes() {
        System.out.println("Titular: " + titular + ", Saldo: R$ " + saldo);
    }
}

public class Main {
    public static void main(String[] args) {
        ContaBancaria conta = new ContaBancaria("João", 1000.00);
        conta.exibirInformacoes();

        try {
            conta.sacar(200.00);
            System.out.println("Saldo após saque: R$ " + conta.getSaldo());

            conta.sacar(900.00);
        } catch (SaldoInsuficienteException e) {
            System.out.println(e.getMessage());
        }
    }
}

File: Lista/Python/Q15.py
class SaldoInsuficienteException(Exception):
    def __init__(self, saldo_atual, valor_saque):
        super().__init__(f'Saldo insuficiente: R$ {saldo_atual:.2f}. Tentativa de saque: R$ {valor_saque:.2f}.')
        self.saldo_atual = saldo_atual
        self.valor_saque = valor_saque

class ContaBancaria:
    def __init__(self, titular, saldo_inicial=0):
        self.titular = titular
        self.saldo = saldo_inicial

    def depositar(self, valor):
        if valor > 0:
            self.saldo += valor
        else:
            raise ValueError("O valor do depósito deve ser positivo.")

    def sacar(self, valor):
        if valor > self.saldo:
            raise SaldoInsuficienteException(self.saldo, valor)
        self.saldo -= valor

    def obter_saldo(self):
        return self.saldo

conta = ContaBancaria('João', 1000)

try:
    conta.sacar(1200)
except SaldoInsuficienteException as e:
    print(e) 

print(f'Saldo atual: R$ {conta.obter_saldo():.2f}')
